src/components/molecules/NavSearchBar/NavSearchBar.tsx

```diff
@@ -1,69 +1,76 @@
-import React, { useCallback, useEffect, useState } from "react";
+import React, { useCallback, useEffect, useState, ChangeEvent } from "react";
+import classNames from "classnames";
+
+import { faSearch } from "@fortawesome/free-solid-svg-icons";
 
 import { VenueEvent } from "types/venues";
 import { Room, RoomTypes } from "types/rooms";
 import { User } from "types/User";
 
 import { WithId } from "utils/id";
-import { currentVenueSelectorData, venueEventsSelector } from "utils/selectors";
 import { isTruthy } from "utils/types";
+import { formatToWeekdayFullname, formatUtcSeconds } from "utils/time";
+import { currentVenueSelectorData, venueEventsSelector } from "utils/selectors";
 
 import { useWorldUsers } from "hooks/users";
 import { useSelector } from "hooks/useSelector";
 import { useProfileModalControls } from "hooks/useProfileModalControls";
 
+import { InputField } from "components/atoms/InputField";
 import { RoomModal } from "components/templates/PartyMap/components";
 
+import { NavSearchResult } from "./NavSearchResult";
+
 import "./NavSearchBar.scss";
-import { InputField } from "components/atoms/InputField";
-import { faSearch } from "@fortawesome/free-solid-svg-icons";
 
-interface SearchResult {
-  rooms: Room[];
-  users: readonly WithId<User>[];
-  events: VenueEvent[];
-}
+type WithImageUrl<T extends object> = T & { image_url?: string };
+
+const emptyEventsArray: VenueEvent[] = [];
 
 const NavSearchBar = () => {
   const [searchQuery, setSearchQuery] = useState("");
   const onSearchInputChange = useCallback(
-    (e) => setSearchQuery(e.target.value),
+    (e: ChangeEvent<HTMLInputElement>) => setSearchQuery(e.target.value),
     []
   );
 
-  const [searchResult, setSearchResult] = useState<SearchResult>({
-    rooms: [],
-    users: [],
-    events: [],
-  });
+  const [foundRooms, setFoundRooms] = useState<Room[]>([]);
+  const [foundUsers, setFoundUsers] = useState<readonly WithId<User>[]>([]);
+  const [foundEvents, setFoundEvents] = useState<WithImageUrl<VenueEvent>[]>(
+    []
+  );
 
   const [selectedRoom, setSelectedRoom] = useState<Room>();
   const hasSelectedRoom = !!selectedRoom;
 
   const venue = useSelector(currentVenueSelectorData);
-  const venueEvents = useSelector(venueEventsSelector) ?? [];
+  const venueEvents = useSelector(venueEventsSelector) ?? emptyEventsArray;
   const { worldUsers } = useWorldUsers();
 
   const { openUserProfileModal } = useProfileModalControls();
 
   useEffect(() => {
     const normalizedSearchQuery = searchQuery.toLowerCase();
     if (!normalizedSearchQuery) {
```

Comment:
create 3 separate useMemos for search results https://github.com/sparkletown/sparkle/pull/1328/commits/1ae5d879a3e476a6ae049ff7058d7218029c4c1a

but now I get Code Climate issue. Need to refactor that. Any ideas about how would be better to split the logic? :thinking:

Reply:
Depending on what code climate is saying, it might be fine to ignore it/etc. It's a guide more than a rule, and it will often have false positives.

Reply:
Looking at the diff you linked, my guess is it's complaining about duplicates based on the `normalizeSearchQuery` function call?

If so, I wonder if we can move it to the 'setter' rather than doing it when we read it back/etc? Alternatively, perhaps we can just move it outside of the memo'd code blocks and have it run all the time. It's probably not too expensive to run in general.

It looks like the `normalizeSearchQuery` function definition might be inside the component render as well, since it seems like it's a static function that doesn't 'close over' any state, I think we could remove that from the component body and have it defined in the file itself.

Reply:
great idea to lowercase string in the 'setter'. 
Doing that in the function `onSearchInputChange` https://github.com/sparkletown/sparkle/pull/1328/commits/c50dba7f0cdf5c19441f06c1edd9889d40e3cb16

`normalizeSearchQuery` is removed. I created this function as I normalized searchQuery 3 times, so in case we would need to change normalization function in future, we would do it only once. But now, as we normalize it onSet, we don't need the function anymore.

btw, a question. I was wondering if we need to trim() search string while searching its matches? or do you think we need to keep whitespaces?

Reply:
Um.. good question.. I would probably suggest looking at the validations for the various things we are searching over and see if they allow whitespace before/after.

Though I suppose if we were to trim both the search query and the items we're comparing against, that probably wouldn't matter..

So yeah, that would probably work fine I think.

Reply:
one minus of trim() is that someone can use whitespace in the beginning or in the end intentionally to find matches at the beginning / at the end of the word (at least I do so sometimes).

Reply:
Happy for you to make a judgement call on this